### PR TITLE
feat: Add support for uploading files as a list

### PR DESCRIPTION
**Pull request by vanhanit**
Uploading a list of files with the same key is supported by the Requests library. This allows the receiving server to accept a list of files for the same key, i.e. accepting any number of files.

Issue: #401

### CODE REVIEW

**Comment by vanhanit** [comment]
This could also be a candidate for a back-port to the 0.9 version as well.

File: atests/http_server/helpers.py [hunk 1]
```diff
@@ -110,15 +110,16 @@ def get_files():
...
-    for k, v in request.files.items():
-        content_type = request.files[k].content_type or 'application/octet-stream'
-        val = json_safe(v.read(), content_type)
-        if files.get(k):
-            if not isinstance(files[k], list):
-                files[k] = [files[k]]
-            files[k].append(val)
-        else:
-            files[k] = val
+    for k in request.files.keys():
```
Comment (lucagiove) [inline atests/http_server/helpers.py:113]: you also had to fix httpbin?

File: src/RequestsLibrary/RequestsKeywords.py [hunk 2]
```diff
@@ -48,9 +49,16 @@ def _common_request(self, method, session, uri, **kwargs):
 
         files = kwargs.get("files", {}) or {}
         data = kwargs.get("data", []) or []
-        files_descriptor_to_close = filter(
-            is_file_descriptor, list(files.values()) + [data]
-        )
+
```
Comment (lucagiove) [inline src/RequestsLibrary/RequestsKeywords.py:52]: at this point I would extract the logic to close file descriptor in a dedicated function and try to cleanup this part

**Review by lucagiove — CHANGES_REQUESTED** [review]
I would also add a unit test to make sure files get closed for both cases list or tuple.
Thanks!

File: atests/http_server/helpers.py (hunk 1 above)
Comment (vanhanit) [reply atests/http_server/helpers.py:113]: Yes, as this only returned one item the way it was written. According to what I found out, you need to use the **getlist(key)** method to get all files for one key.

File: src/RequestsLibrary/RequestsKeywords.py (hunk 2 above)
Comment (vanhanit) [reply src/RequestsLibrary/RequestsKeywords.py:52]: I refactored the closing of file descriptors into its own static method.

**Comment by vanhanit** [comment]
Added verification to the Robot Framework tests to verify that the file descriptors are closed for both use cases.

**Comment by lucagiove** [comment]
> Added verification to the Robot Framework tests to verify that the file descriptors are closed for both use cases.

there was already a unit test for check file closing

File: src/RequestsLibrary/RequestsKeywords.py [hunk 3]
```diff
@@ -176,7 +192,7 @@ def session_less_get(
...
         | ``cookies``  | Dict or CookieJar object to send with the request. |
-        | ``files``    | Dictionary of file-like-objects (or ``{'name': file-tuple}``) for multipart encoding upload. ``file-tuple`` can be a 2-tuple ``('filename', fileobj)``, 3-tuple ``('filename', fileobj, 'content_type')`` or a 4-tuple ``('filename', fileobj, 'content_type', custom_headers)``, where ``'content-type'`` is a string defining the content type of the given file and ``custom_headers`` a dict-like object containing additional headers to add for the file. |
+        | ``files``    | Dictionary of file-like-objects (or ``{'name': file-tuple}``) for multipart encoding upload. ``file-tuple`` can be a 2-tuple ``('filename', fileobj)``, 3-tuple ``('filename', fileobj, 'content_type')`` or a 4-tuple ``('filename', fileobj, 'content_type', custom_headers)``, where ``'content-type'`` is a string defining the content type of the given file and ``custom_headers`` a dict-like object containing additional headers to add for the file. List or tuple of ``('key': file-tuple)`` allows uploading multiple files with the same key, resulting in a list of files on the receiving end. |
```
Comment (lucagiove) [inline src/RequestsLibrary/RequestsKeywords.py:195]: the strange thing is that even the official documentation don't mention files as list or tuple..
https://requests.readthedocs.io/en/latest/api/#requests.request

Comment (vanhanit) [reply src/RequestsLibrary/RequestsKeywords.py:195]: Yes, I know that they don't mention it. But the support is there and I don't know how long it has been there.

**Comment by vanhanit** [comment]
> > Added verification to the Robot Framework tests to verify that the file descriptors are closed for both use cases.
> 
> there was already a unit test for check file closing

Well, I did not dig too deep to find that.
The Robot test ensures that both a dictionary and a list is tested.
It does not test having files in the data section. Maybe the unit test covers this?

**Comment by vanhanit** [comment]
I rebased on master. I don't think that my changes is what is causing the test to fail.

**Comment by lucagiove** [comment]
> I rebased on master. I don't think that my changes is what is causing the test to fail.

no it's not your commit, there are still some tests that use a web version of httpbin, sometimes they throttle the requests...
they should be changed but there are tests also for https certificates so not easy to implement locally

**Comment by vanhanit** [comment]
> > I rebased on master. I don't think that my changes is what is causing the test to fail.
> 
> no it's not your commit, there are still some tests that use a web version of httpbin, sometimes they throttle the requests... they should be changed but there are tests also for https certificates so not easy to implement locally

Yeah, it is the same test that also fails locally for me.

**Comment by vanhanit** [comment]
@lucagiove Was there anything I still needed to do with this? How do we rerun the builds?

**Review by lucagiove — APPROVED** [review]
No I would have added also the unit test but mandatory.
I'll try to have the pipeline running.

**Comment by lucagiove** [comment]
> > No I would have added also the unit test but mandatory. I'll try to have the pipeline running.
> 
> But didn't you say that there already are a unit test for the closing of file descriptors? Or was there some other unit test that you wanted to run? Either way, there are a couple of Robot Framework tests that cover the functionality.

Yep but not with the specific input of a list.

**Pull request merged**